File: 89_k-sum/k-sum.cpp
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/k-sum
@Language: C++
@Datetime: 14-12-23 03:03
*/

class Solution {
public:
    /**
     * @param A: an integer array.
     * @param k: a positive integer (k <= length(A))
     * @param target: a integer
     * @return an integer
     */
    int kSum(vector<int> A,int k,int target) {
        // wirte your code here
        vector<vector<vector<int>>> d(A.size()+1,vector<vector<int>>(target+1,vector<int>(k+1,0)));
        for(int i=0;i<=A.size();++i)
            d[i][0][0]=1;
        for(int i=1;i<=A.size();++i)
            for(int j=1;j<=target;++j)
                for(int m=1; m<=min(k,i); ++m)
                    d[i][j][m] = d[i-1][j][m] + (j >= A[i - 1] ? d[i-1][j-A[i-1]][m-1] : 0);
        return d[A.size()][target][k];
    }
};

File: 89_k-sum/k-sum.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/k-sum
@Language: Python
@Datetime: 14-12-23 01:05
'''

class Solution:
    """
    @param A: An integer array.
    @param k: a positive integer (k <= length(A))
    @param target: integer
    @return an integer
    """
    def kSum(self, A, k, target):
        # write your code here
        d = [[0]*(k+1) for i in xrange (target + 1)]
        d[0][0] = 1
        for i in xrange (1, len(A) + 1):
            for j in xrange (target, A[i-1]-1, -1):
                for  l in xrange (k, 0, -1):
                    d[j][l] = d[j][l] + d[j-A[i-1]][l-1]
        return d[target][k]